lib/fauna_model/base.rb

```
@@ -0,0 +1,269 @@
+        @deleted
+      end
+
+      def persisted?
+        !new_record? && Model.calculate_diff(@original, @current).empty?
+      end
+
+      def id
+        ref.value.split('/').last unless ref.nil?
+      end
+
+      def save(validate = true)
+        return false if validate && invalid?
+
+        if new_record?
+          self.class.from_fauna(Fauna::Context.query(create_query))
+        else
+          self.class.from_fauna(Fauna::Context.query(update_query))
+        end
+      end
+
+      def save!(validate = true)
+        fail InvalidInstance.new('Invalid instance data') if validate && invalid?
+
+        old_changes = changes
+
+        if new_record?
+          result = from_resource(Fauna::Context.query(create_query))
+        else
+          result = from_resource(Fauna::Context.query(update_query))
+        end
+
+        @previous_changes = old_changes
+
+        result
+      end
+
+      def update(params = {})
+        model = copy
+        model.send(:apply_params, params)
+        model.save
+      end
+
+      def update!(params = {})
+        apply_params(params)
+        save!
+      end
+
+      def destroy
+        return false if new_record?
+        Fauna::Context.query(delete_query)
+        @deleted = true
+      end
+
+    private
+
+      def create_query
+        return nil unless new_record?
+        Fauna::Query.create(self.class.fauna_class, Fauna::Query.quote(query_params))
+      end
+
+      def replace_query
+        return nil if new_record?
+        Fauna::Query.replace(@current['ref'], Fauna::Query.quote(query_params))
+      end
+
+      def update_query
+        return nil if new_record?
+        Fauna::Query.update(@current['ref'], Fauna::Query.quote(Model.calculate_diff(@original, @current)))
+      end
+
+      def delete_query
+        return nil if new_record?
+        Fauna::Query.delete(ref)
+      end
+
+      def query_params
+        params = Model.hash_dup(@current)
+
+        # Remove unsettable fields
```

Do setters still exist for these fields? If so it's a kind of silent failure.